Mlem/App/Globals/Definitions/AppState.swift

```diff
@@ -8,120 +8,78 @@
-        let newAccount = ActiveAccount(account: account)
-        activeAccounts.forEach { $0.deactivate() }
-        guestAccount.deactivate()
-        activeAccounts = [newAccount]
-    }
+    private(set) var guestSession: GuestSession!
+    private(set) var activeSessions: [UserSession] = []
     
-    func enterGuestMode(for instanceUrl: URL) {
-        activeAccounts.forEach { $0.deactivate() }
-        activeAccounts = []
-        guestAccount.deactivate()
-        guestAccount = .init(instanceUrl: instanceUrl)
+    private init() {
+        self.guestSession = .init(account: AccountsTracker.main.defaultGuestAccount)
+        changeAccount(to: AccountsTracker.main.defaultAccount, deactivateOldGuest: false)
     }
     
-    func deactivate(account: Account) {
-        if let index = AppState.main.activeAccounts.firstIndex(where: { $0.account === account }) {
-            activeAccounts[index].deactivate()
-            activeAccounts.remove(at: index)
-            if activeAccounts.isEmpty, let defaultAccount = AccountsTracker.main.defaultAccount {
-                changeUser(to: defaultAccount)
-            }
-        }
+    func changeAccount(to account: any Account) {
+        changeAccount(to: account, deactivateOldGuest: true)
     }
     
-    func enterOnboarding() {
-        activeAccounts.removeAll()
+    private func changeAccount(to account: any Account, deactivateOldGuest: Bool) {
+        ToastModel.main.add(.account(account))
+        
+        activeSessions.forEach { $0.deactivate() }
+        if deactivateOldGuest {
+            guestSession?.deactivate()
+        }
+        
+        // Save because we updated `lastUsed` in the above `deactivate()` calls
+        AccountsTracker.main.saveAccounts(ofType: .all)
+        
+        if let account = account as? UserAccount {
+            let activeAccount = UserSession(account: account)
+            activeSessions = [activeAccount]
+        } else if let account = account as? GuestAccount {
+            activeSessions = []
+            guestSession = .init(account: account)
+            GuestAccountCache.main.clean()
+        } else {
+            assertionFailure()
+        }
     }
     
-    private init() {
-        if let user = AccountsTracker.main.defaultAccount {
-            changeUser(to: user)
-        } else if let user = AccountsTracker.main.savedAccounts.first {
-            changeUser(to: user)
+    func deactivate(account: any Account) {
+        if let account = account as? UserAccount {
+            if let index = AppState.main.activeSessions.firstIndex(where: { $0.account === account }) {
+                activeSessions[index].deactivate()
+                activeSessions.remove(at: index)
+            }
+        } else if let account = account as? GuestAccount {
+            guard account == guestSession.account else {
+                assertionFailure("Tried to deactivate a non-active GuestAccount")
+                return
```

Comment:
This prevents saved guest accounts from being removed if they're not active--is this intentional?

Reply:
Nice catch 👍